Privacy policy

NoahMcGraw opened the pull request:
Phase 2- Add the rest of the questionnaire and the summary page. Policy can now be completely edited and generated/printed. 

Notes: 
Discussed printing method did not work as expected. Attempted to create a workaround but ran into issues.. Position: fixed would not display the full contents of the div on print. position: absolute Did display the contents of the div but could not gain full use of the window Because it was the child of a div that was separated from the banner and rest of the app and therefore could only, at the most, occupy the full dimensions of the parent div. Would not be an issue since we could add .:not(.generatedPolicy) to set everything else to display: none, EXCEPT that this also applies to the children of generatedPolicy and specifying .generatedPolicy > * in the :not does not seem to be effective.

Regardless, I was actually able to get the previous print function to work across browsers(Including Chrome) with relative ease AND I think it is a much cleaner way to do this anyway so I will include this version in the push.

A phone number mask has been added. However, the mask does not work very well with ng-input and will clash with the labels that already exist.

NoahMcGraw committed:
Changed desciption inut in tabs to md-chips; Began work on finishedPolicy
partial;

NoahMcGraw committed:
Finished summary page for generated policy; Changed email input in form to
phone number for relevance; Added print functionality but cannot retain css
at the moment, needs work;

NoahMcGraw committed:
Added printElem function to format doc and send it to printer;
Added formatPhoneNumber function to format the phone number given;
Added generatedPolicy directive;
md-chips click box is not scaling, needs work;

NoahMcGraw committed:
Changed placeholders in md-chips to give examples of useful clauses;

NoahMcGraw committed:
Add disclaimer for "Not Legal Advice";
add md-add-on-blur="true" to the chips;
Update Angular to 1.1.10;
Add mask for phone number to input;
Add input for company compliance email;
Fix print function to work across major browsers;
Change md-chip placeholders to shorter examples;

NoahMcGraw committed:
Add printing var change in printElem function;
Remove formatPhoneNumber function;
Remove printGeneratedPolicy function;

reesemartin commented on app/controllers/main-ctrl.js:
```diff
@@ -64,4 +82,53 @@ app.controller('mainCtrl', ['$scope', '$mdDialog', function($scope, $mdDialog){
+        var head = "";
+        var body = "";
+        var title = "<title>" + $scope.newPolicy.companyName + " Privacy Policy</title>";
+        var angularMaterialCss = '<link rel="stylesheet" href="https://ajax.googleapis.com/ajax/libs/angular_material/1.1.8/angular-material.min.css">';
```
Might also be a good idea to include the app.css to ensure any global default styles of your app are included.

reesemartin commented on app/partials/privacyPolicyGenerator-template.html:
```diff
@@ -9,17 +9,18 @@
+            <md-button class="md-primary md-raised unisyn-big-button" ng-click="startNewPolicy()" ng-show="newPolicy.untouched">Let's Get Started</md-button>
         </md-content>
     </div>
     <div layout="column" layout-align="center center">
```
Here is one the clear issues on mobile.

layout directives should not be used when you are using pixels to restrict its children. The point of the directive is that you make it into a "table" and the "cells" are portions that add up to 100%. A rare exception would be if you need vertical centering as well. But here you only appear to need horizontal centering.

If all you care about is that its content is centered then move the privacyPolicyGeneratorContainer class to this element as the actual policy generator contain and then set its width as described in the overall PR review and then give it:
```
margin: auto;
```
which will center horizontally.

reesemartin commented on app/partials/questionnaireTabs/questionnaireUsage-template.html:
```diff
@@ -51,12 +51,19 @@
+<div ng-show="questionnaire.usage.answer" class="md-vertical-padding-bottom">
+    <div class="md-title">Please list the additional reasons you will access user data</div>
+    <md-chips ng-model="questionnaire.usage.answerDescription"
+              readonly="false"
```
This doesnt need to be readonly="false" since false is the default if you dont even include this attribute.

reesemartin commented on app/partials/questionnaireTabs/questionnaireStorage-template.html:
```diff
@@ -59,10 +59,19 @@
+        Please list any additional data that will be tracked and stored
+    </div>
+    <md-chips ng-model="questionnaire.storage.answerDescription"
+              readonly="false"
```
This doesnt need to be readonly="false" since false is the default if you dont even include this attribute.

reesemartin commented on app/partials/questionnaireTabs/questionnaireSharing-template.html:
```diff
@@ -28,23 +28,32 @@
+        Please list the additional reasons you will share user data
+    </div>
+    <md-chips ng-model="questionnaire.sharing.answerDescription"
+              readonly="false"
```
This doesnt need to be readonly="false" since false is the default if you dont even include this attribute.

reesemartin reviewed (changes requested):
The app is not very mobile friendly. This is mostly due to the px widths you have defined. See the stylesheet notes I have added to our coding guidelines repo here: https://github.com/UniSynTechnologies/coding-guidelines/blob/master/stylesheets.
Instead of saying width 900px make it width 100% with max-width of 900px. Though I dont think that is even necessary. Set a max width on the app itself and then let all the contents be 100%. That makes your app more modular so that if you drop this into something else, you can control its width / mobile aspects based on the top level vs each thing inside trying to be its own width.

Also with styling I notice you have some scroll bars on mobile due to the padding not being included when doing widths. This is due to the box-sizing css property which most apps set as "border-box". We do this in our reset.css file in TopSale. You can just put a quick rule at the top of app.css like so:
```
privacy-policy-generator * {
    box-sizing: border-box;
}
```

It is a generally a good idea to adjust the width of your browser window whenever you make css changes to make sure they look good on mobile. I try to do a mobile first design methodology where I design the interface with my browser window thin like mobile and then I expand it and add conditions to spread things out for desktop.

This is a good way to catch mobile issues like the ones I described early on before it requires hacking out parts of your code to fix layout.

-------------------------------------------------

The issue with your printing of the policy fullscreen was likely due to the 600px height you have on the policy container and the zindex. The following rule applied to generatedPolicy does work in my dev:
```
    position: fixed;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: #fff;
    z-index: 999999;
    height: initial;
```
But I do see what you have in place and it works too. I dont like that there is html in the controller since the breaks the model / view / controller scheme but its not worth spending the time to change it since it works pretty well. Great job confirming it works in other browsers.

----------------------------------------------------

Final nitpick thing:
All files should have 1 blank line at the end of the file. Lots of discussion about this general "rule" here: https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline but it really comes down to being because some text parsers only could lines ending in new line characters are actual lines to parse. So if our code is ever passed through a 3rd party parser for minification or uglification or some other automated process there is a slight chance this could be a problem.

Look at the files changed in this PR and you will see that git actually makes a red notation at the end of your files for the missing new line character.

Try to remember to add new lines at the end of the files. I know its not easy to remember but try your best to.

NoahMcGraw committed:
Change sizing for privacyPolicyGeneratorContainer to be more mobile friendly
with max-widths and margin: auto; Add new line to the end of all files in app;
Add app.css to generated policy print page; Remove readonly attribute
from md-chips; Add blank placeholder to phone number input to avoid
letting the mask and the label overlap when unfocused and empty;
Add nav buttons for mobile version using md-icons and change layout where
neccesary; Allow swipe actions on md-chips; Change disclaimer css to be
mobile friendly;
